platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbCompactionCoordinator.java

```diff
@@ -0,0 +1,282 @@
+            if (compactionFuturesByName.containsKey(task.id)) {
+                CompletableFuture<?> completableFuture =
+                        compactionFuturesByName.get(task.id).asCompletableFuture();
+                if (completableFuture.isDone()) {
+                    compactionFuturesByName.remove(task.id);
+                } else {
+                    logger.debug(MERKLE_DB.getMarker(), "Compaction for {} is already in progress", task.id);
+                    return;
+                }
+            }
+
+            InterruptibleCompletableFuture<Boolean> future =
+                    InterruptibleCompletableFuture.runAsyncInterruptibly(task, executor);
+            future.asCompletableFuture().thenAccept(result -> {
+                if (result) {
+                    statisticsUpdater.updateStoreFileStats();
+                    statisticsUpdater.updateOffHeapStats();
+                    logger.info("Finished compaction for " + task.id);
+                }
+            });
+            compactionFuturesByName.put(task.id, future);
+        }
+    }
+
+    /**
+     * Creates a thread pool for compaction tasks if it does not exist yet, or returns the existing one.
+     * If the existing thread pool is shutdown, it will be replaced with a new one. Note that there is only one
+     * compacting thread-pool per JVM.
+     * @return a thread pool for compaction tasks
+     */
+    ExecutorService createOrGetCompactingExecutor() {
+        ExecutorService executorService = compactionExecutorServiceRef.get();
+        if (executorService == null) {
+            executorService = new ThreadPoolExecutor(
+                    getCompactingThreadNumber(),
+                    getCompactingThreadNumber(),
+                    0L,
+                    TimeUnit.MILLISECONDS,
+                    new LinkedBlockingDeque<>(),
+                    new ThreadConfiguration(getStaticThreadManager())
+                            .setThreadGroup(new ThreadGroup("Compaction"))
+                            .setComponent(MERKLEDB_COMPONENT)
+                            .setThreadName("Compacting")
+                            .setExceptionHandler((t, ex) ->
+                                    logger.error(EXCEPTION.getMarker(), "Uncaught exception during merging", ex))
+                            .buildFactory());
+            if (!compactionExecutorServiceRef.compareAndSet(null, executorService)) {
+                try {
+                    executorService.shutdown();
+                } catch (Exception e) {
+                    logger.error(EXCEPTION.getMarker(), "Failed to shutdown compaction executor service", e);
+                }
+            }
+        } else {
+            if (executorService.isShutdown()) {
+                compactionExecutorServiceRef.compareAndSet(executorService, null);
+                return createOrGetCompactingExecutor();
+            }
+        }
+
+        return compactionExecutorServiceRef.get();
```

Comment:
I think this could be written a bit more safely and slightly simplified while still avoiding a full lock (I assume that's a goal here).
Possible suggestion:
```suggestion
        if (compactionExecutorServiceRef.getAcquire() == null) {
            compactionExecutorServiceRef.compareAndSet(null,
                    new ThreadPoolExecutor(
                            getCompactingThreadNumber(),
                            getCompactingThreadNumber(),
                            0L,
                            TimeUnit.MILLISECONDS,
                            new LinkedBlockingQueue<>(MAX_TABLES),
                            new ThreadConfiguration(getStaticThreadManager())
                                    .setThreadGroup(new ThreadGroup("Compaction"))
                                    .setComponent(MERKLEDB_COMPONENT)
                                    .setThreadName("Compacting")
                                    .setExceptionHandler((t, ex) ->
                                            logger.error(EXCEPTION.getMarker(), "Uncaught exception during merging", ex))
                                    .buildFactory()));
        } else {
            ExecutorService executorService = compactionExecutorServiceRef.get();
            if (executorService.isShutdown()) {
                compactionExecutorServiceRef.compareAndSet(executorService, null);
                return createOrGetCompactingExecutor();
            }
        }
        return compactionExecutorServiceRef.getAcquire();
```

Reply:
Interesting. A couple questions: 
- could you please explain why you use `getAquire`? If understand correctly it has the same memory semantics as acquiring  a lock. I'd guess it's slightly more lightweight operation than volatile read but it provides what we need here. Is my understanding correct? 
- Theoretically there may be a data race, and two threads will create two instances of `ThreadPoolExecutor`, and one of them will be effectively discarded. Shouldn't we shut it down if it's not used?

Reply:
The getAcquire is, indeed, a little lighter than volatile read, but it's not quite the same as acquiring a lock.  The actual formal definition of the memory orders is [here](https://en.cppreference.com/w/cpp/atomic/memory_order).

I hadn't thought of the created executor, that's a slight added wrinkle...  We could still do something very similar, but we do probably need to handle that condition.

Side note, if we think this might be called more than once, we could make it easier to read by creating a constant for the ThreadConfiguration instead of creating a new one every time...

Reply:
>. We could still do something very similar, but we do probably need to handle that condition.

I would be really interested to see how exactly it without locks and without creating a redundant instance.